### ISSUE TITLE
Translate.py - hard to detect convergence

### ISSUE THREAD

**Issue by anjishnu** [tensorflow/tensorflow#550]
In the Sequence to Sequence machine translation example - the training pipeline only outputs perplexity numbers which are hard to interpret. If it could also output some sample translations it would be much easier for people to measure convergence

Anecdotally perplexities of around ~5 result in horrible translations - has anyone gotten it to work, and if so, what perplexities did you need to reach before you started getting reasonable translations?

**Comment by juesato** [comment]
I'm having the same experience - after about 9 hours of training and 15000 steps, the model has reached perplexity ~10, but the predictions are almost all _UNK and even after squashing all those probabilities to 0, the translations are really poor (with a tendency to output just a few random words when translating short sentences).
As a silver lining (depending on how you look at it), it does get the example "Who is the president of the United States?" right.

**Comment by marcotrombetti** [comment]
@anjishnu I agree and I also would like to recommend showing a few examples translation for convergence or BLUE scores. A log file would be perfect.

@juesato I think that there is a bug somewhere in the tutorial. I tried to make it work with a small subset and different configurations but I never got a plausible translation. I am trying to debug it. Until now now errors on the data and the script until the seq2seq model creation. Continuing...

See my issue #600

**Comment by felipemoraes** [comment]
Same here I trained during two days and nothing. Have you solve the problem @marcotrombetti ?

**Comment by lukaszkaiser** [comment]
To get reasonable translations you need to train a reasonably large network (say 3 layers with 1K nodes) for at least a few hundred thousand steps. Indeed, the perplexity numbers are just a guide to what's happening, but you can get decodings from the binary as well (or even add your own BLEU score computation using the same function). Did you try to re-run the tutorial setting? That was one decoding I got when training the tutorial before first release:

```
Reading model parameters from /tmp/translate.ckpt-340000
>  Who is the president of the United States?
 Qui est le président des États-Unis 
```

We should be able to get at least something like this, if not, there might be some new bug to correct. I'll start a run to check it, but it usually takes about a week to get to this point, so please, post here if you get some numbers earlier.

**Comment by lukaszkaiser** [comment]
I ran the tutorial model over the weekend + 1 day on a K40. It got to 193K steps (3 layers by 1024) and to reasonable perplexity and samples.

```
global step 193000 learning rate 0.0861 step-time 1.70 perplexity 4.17
  eval: bucket 0 perplexity 5.02
  eval: bucket 1 perplexity 3.27
  eval: bucket 2 perplexity 3.51
  eval: bucket 3 perplexity 4.82

Reading model parameters from /tmp/translate.ckpt-193000
> Who is the president of the United States?
Qui est le président des États-Unis ?
```

So I think it works as expected. Better results can be obtained if you tune the model, hopefully also after we close #640 or with more training time. But what it produces looks reasonable to me, so I'm closing this issue. It would be great if someone could confirm what I see -- please re-open if it does not work for you.

**Comment by anjishnu** [comment]
I'm thinking we can solve this once and for all by simply printing some sample outputs at every checkpoint so it's easy people to see if it's starting to make sense?

**Comment by lukaszkaiser** [comment]
I think it's not too hard now, just look for perplexity to go down to around 4, maybe 4.2 -- and the lower the better, esp. on more complex sentences. But indeed, it's a bit of a magic number and some examples might be good to have. Code contributions are welcome, just send your CL!

**Comment by markusdr** [comment]
It would also be helpful to output BLEU scores on the dev data set at each checkpoint.

Someone should train on the same data as this [Moses baseline](http://www.statmt.org/moses/?n=moses.baseline) and compare the final BLEU scores when decoding the test set (23.5 with Moses). That would be a great comparison point and sanity check; it should also train fast as the data is not that large. (Use the same tokenizer as given there instead of the tutorial tokenizer in data_utils.py). 
I've tried it myself, but a model of size 1024 with 3 layers doesn't fit in my 3GB GPU memory, and I haven't figured out yet how to get reasonable performance with a smaller model.

**Comment by juesato** [comment]
The basically problem I'm observing is that while the model learns to translate "Who is the president of the United States?" it doesn't translate other very basic sentences well. Even though I had perplexity <5, basic translations such as "What is my name?" and "How are you doing today?" would not be translated well, and often have many OOV tokens (I haven't limited max vocab size at all). My concern is if perplexity isn't a good measure, for example, the model can game the cost function by frequently predicting OOV.

I would post samples, but I'm out of the country without my computer - I'll be able to do this in a week.

@lukaszkaiser could you post a few more sample translations? If your model seems robust, then maybe it's just a peculiarity on my end, but right now, I'm not convinced perplexity <5 means much. 

Thank you for looking into this, really appreciate it.

**Comment by lukaszkaiser** [comment]
One thing that needs to be remembered is that neural translation is a young and active area of reseach -- it's not like there are established ways for everything. Let me comment on two points.

(1) Perplexity vs BLEU. In a number of recent papers on neural translation many researchers report that perplexity correlates well with BLEU scores. Some papers even claim that perplexity correlates better with human judgement of translations than BLEU. But these are all just imperfect measures, sometimes imperfect in different ways. Looking at a few self-chosen samples is also an imperfect measure: the ones you consider might just be more outside of the training data (which comes from newspapers if I understand right), but not really harder. So in the end one needs to look at all of them to get a reasonable measurement and think hard what they mean.

(2) OOV words. This is a well-known problem in neural translation models and a there are ongoing research efforts to deal with it. One way that seems to work well is having multiple UNK tokens and doing some replacements, see this paper: http://arxiv.org/abs/1410.8206 . In a basic model like the one in the tutorial the UNK is heavily over-represented in the training data which leads to bad decodings and a bit "cheated" perplexity, as you said. You can try to just decrease UNK probability in the training by using a few different UNKs, or the method from the above paper, or something else entirely.

I'm writing about these points to illustrate one thing: this is an ongoing research effort. We constructed the tutorial to show how to get started with these kinds of models in TensorFlow, but all these research questions remain open. You can now see where the problems are and start working on your own solutions -- that's the intention!

**Comment by juesato** [comment]
@lucaszkaiser Thanks for the response - I think that's a fair answer, and the explanations of these two challenges is helpful. I'm not sure I'll come back to this project, since there's other projects I'm working on, but I'll try digging deeper into your suggestions if I do.

Your response clarified a lot for me - I'd suggest adding something along those lines to the tutorial page, or at least a mention along the lines of "neural translation is still quickly improving and models are flawed - this model is only a starting point." I think it is easy to get the impression, as I and I'll venture others on this issue, based on the writing of the tutorial and the sample translation, that the model should just work out of the box.

**Comment by lukaszkaiser** [comment]
Just for completeness, I ran the queries you suggested through my model.

```
> What is your name ?
Quelle est votre nom ?
> What is my name ?
Qu ' est-ce que mon nom ?
> How are you doing today ?
Comment _UNK aujourd’hui ?
```

So indeed - there is a strange _UNK there, but it's not bad, right? In fact quite funny in some ways - I think everyday queries are really a bit outside of the training data used there.

**Comment by juesato** [comment]
Hm those results definitely look better than what I remember - I'll take a look back at my model and update this for completeness in a week.

Point regarding everyday questions absent in the training data is well-taken.

**Comment by anjishnu** [comment]
> Code contributions are welcome, just send your CL!

Haha, I'd love to. Given your CLA requirements I expect to be stuck in corporate purgatory for about 6-8 months before being denied permission. :)

**Comment by lukaszkaiser** [comment]
We made our process for accepting contributions as quick and painless as we could, and there are already many developers submitting their CLs. I suggest you try, and let me know personally if you have any trouble -- we want to make it as easy as possible for anyone to contribute!

**Comment by anjishnu** [comment]
haha, no, I was referring to corporate policies at my employer, not of this project. I've seen the improvement since it began - it's quite impressive!

**Comment by aliabbasjp** [comment]
@lukaszkaiser 
I have the following similar problem, suddenly my perplexity scores shot up from 3 digit numbers to 5+ digit numbers:

http://stackoverflow.com/questions/35126954/how-can-i-know-if-the-epoch-point-is-reached-in-seq2seq-model

**Comment by markusdr** [comment]
@aliabbasjp It looks like your dev data is very different from your training data. Maybe there are many unknown words given the vocabulary size you picked. Try running with dev data that is closer to your training data, e.g., split 10% off your current training data and use it as dev data.

**Comment by aliabbasjp** [comment]
@markusdr No the training data is of the same domain and has similar words and I am not doing a translation but both pairs of sentences are english sentences.

**Comment by lukaszkaiser** [comment]
I think the comment about the difference between your dev and train data came from your log output like this:

```
global step 374600 learning rate 0.0069 step-time 1.92 perplexity 1.02
  eval: bucket 0 perplexity 137268.32
```

Your training perplexity is 1.02 -- the model is basically perfect on the data it receives for training. But your dev perplexity is enormous, the model does not work at all. How did it look in earlier epochs? I would suspect that there is some mismatch in something. Maybe the tokenization is different for train and dev? Maybe the sizes of the buckets from the original translation model are not appropriate for your use-case? If you share more details, I'll be happy to take a look and try to help (maybe better on the original bug).

**Comment by suraj1990** [comment]
Some of the sampel outputs that I am getting are attached 
![sample-translations](https://cloud.githubusercontent.com/assets/5551707/16106680/15101370-33b4-11e6-8cb0-c206469f4141.png).

The results are not good. 
`global step 46300 learning rate 0.2499 step-time 0.89 perplexity 7.50
eval: bucket 0 perplexity 6.22
eval: bucket 1 perplexity 7.30
eval: bucket 2 perplexity 9.18
eval: bucket 3 perplexity 11.42`

Should I train till perplexity goes further down?  and for the sentences in the development data, I see that for what ever input the translation is _UNK. what could be the reason for this?

**Comment by lukaszkaiser** [comment]
Yes, train more. From the seq2seq tutorial: "one epoch (going through the training data once) takes about 340K steps with batch-size of 64. At this point the model can be used for translating English sentences to French". You're at step 46K, that's not even 15% of the 340K described! (The UNK problem can be better remedied by improving tokenization, but when you train longer it becomes less of a problem in all cases.)

**Comment by fangyihao** [comment]
With all default params, the result is not good, even at the 335800 steps
![image](https://cloud.githubusercontent.com/assets/22711053/20027603/8aeb77f2-a2ee-11e6-8009-64951300c8b1.png)

**Comment by lukaszkaiser** [comment]
This is very strange, it does not happen to me. Also, your perplexity is very low, 2.15 is suspicious. It looks like you're maybe not using all training data, or maybe the data isn't tokenized right? These perplexities look way too low for the default setting.

**Comment by fangyihao** [comment]
However, what happened is that I downloaded all the python source code from the tensorflow git project without any modification. The tokenizer is also the original basic_tokenizer in data_utils.py. 

I ran the command below as instructed on https://www.tensorflow.org/versions/r0.11/tutorials/seq2seq/index.html. 
I believe I exactly followed all the instructions in the tutorial, but I still cannot reproduce the results. Did I miss anything? 

```
python translate.py
  --data_dir [your_data_directory] --train_dir [checkpoints_directory]
  --en_vocab_size=40000 --fr_vocab_size=40000
```

**Comment by cshapeshifter** [comment]
I have the same problem; haven't found a solution yet. I opened a bug report in tensorflow/models where this example code now resides: https://github.com/tensorflow/models/issues/771

**Comment by julien888** [comment]
I think that I may have a bug in the data_utils.py:
`  if not normalize_digits:
    return [vocabulary.get(w.decode(), UNK_ID) for w in words]
  # Normalize digits by 0 before looking words up in the vocabulary.
  return [vocabulary.get(_DIGIT_RE.sub(b"0", w).decode(), UNK_ID) for w in words]
`

w is a binary string and the vocabulary.get() would ALWAYS fail and result in an UNK. Adding w.decode() to convert it to a string seems to work. I am retraining the model right now, I get some gibberish, but at least no UNK. I am only at step 1500 so it will take a while!

**Comment by julien888** [comment]
I think I may have found a bug in in data_utils.py:
`  if not normalize_digits:
    return [vocabulary.get(w.decode(), UNK_ID) for w in words]

  return [vocabulary.get(_DIGIT_RE.sub(b"0", w).decode(), UNK_ID) for w in words]`

Using python 3.5.2, vocabulary.get(w) does not work when w is a binary string. vocabulary.get(w.decode()) works. I am retraining, at least I do not get the dreaded UNK...

**Comment by cshapeshifter** [comment]
@julien888: Yes, I reported the same error and it's already been fixed:
https://github.com/tensorflow/models/issues/771#issuecomment-269059667

**Comment by sfincke** [comment]
Recently got TensorFlow 1.0 and pulled recent copies of code and data. Old problem with getting mostly _UNK have resolved. However, perplexity is stuck around 9.5:

global step 846400 learning rate 0.0001 step-time 0.55 perplexity 9.49

Decode output bears some resemblance to French, but it's quite off:

> Who is the president of the United States?
Le Conseil de l’Europe a adopté le rapport de M . _UNK .
> The president will return on Thursday.
Le gouvernement du Canada a publié un rapport sur les questions .

**Comment by blester125** [comment]
I am very confused about my results. My network has only trained for 56,600 steps but my perplexities are far lower than many that are reported here. For example by bucket perplexities are 3.92, 2.77, 2.97, 3.91 and the total perplexity is something around 2.15

I started the model training with
python translate.py
  --data_dir [your_data_directory] --train_dir [checkpoints_directory]
  --en_vocab_size=40000 --fr_vocab_size=40000

**Comment by vincentracine** [comment]
@lukaszkaiser do you still have the 193K checkpoint file? (/tmp/translate.ckpt-193000).

**Comment by lukaszkaiser** [comment]
Sorry, I don't have it any more.

**Comment by vincentracine** [comment]
@lukaszkaiser No worries. Should i expect to get OK translations using 2 layers (256 units each) with enough training steps? Or is it a waste of time and I should restart training with the default 3 layers (1024 units each)?

**Comment by lukaszkaiser** [comment]
With the old seq2seq code, I'd say 2 layers of 256 will just not work. 3 layers of 1024 should be fine. But you should really check out the google/seq2seq repo (https://github.com/google/seq2seq). It's a newer and much improved version of TensorFlow seq2seq models. It's may look a bit more complex than the tutorial at first, but if you're after quality and better-tuned models, it'll be worth your while :).

**Comment by vincentracine** [comment]
@lukaszkaiser Awesome suggestion thank you!

**Comment by ngovinhtn** [comment]
Hi all,
i run trainning with translate.py follow to https://www.tensorflow.org/tutorials/seq2seq with 3 layer of 512 units, but perplexity alway increase follow as:
global step 104400 learning rate 0.2375 step-time 1.01 perplexity 1.00
  eval: bucket 0 perplexity 52838.74
  eval: bucket 1 perplexity 117227.88
  eval: bucket 2 perplexity 445883.28
  eval: bucket 3 perplexity 323676.55
global step 104600 learning rate 0.2375 step-time 1.03 perplexity 1.01
  eval: bucket 0 perplexity 36887.39
  eval: bucket 1 perplexity 826537.03
  eval: bucket 2 perplexity 525378.90
  eval: bucket 3 perplexity 304065.98
global step 104800 learning rate 0.2352 step-time 0.95 perplexity 1.00
  eval: bucket 0 perplexity 27844.06
  eval: bucket 1 perplexity 248171.42
  eval: bucket 2 perplexity 550592.36
  eval: bucket 3 perplexity 435554.41
I do not know why? Please help me, I can not bug this problem.
